immortal-cravings.lua

```diff
@@ -0,0 +1,231 @@
+end
+
+---create Eat job for the given unit
+---@param unit df.unit
+local function goEat(unit)
+    local meal = get_closest_meal(unit.pos)
+    if not meal then
+        -- print('no accessible meals found')
+        return
+    end
+    local job = idle.make_job()
+    job.job_type = df.job_type.Eat
+    job.flags.special = true
+    local dx, dy, dz = dfhack.items.getPosition(meal)
+    job.pos = xyz2pos(dx, dy, dz)
+    if not dfhack.job.attachJobItem(job, meal, df.job_item_ref.T_role.Other, -1, -1) then
+        error('could not attach meal')
+        return
+    end
+    dfhack.job.addWorker(job, unit)
+    local name = dfhack.TranslateName(dfhack.units.getVisibleName(unit))
+    print(dfhack.df2console('immortal-cravings: %s is getting something to eat'):format(name))
+end
+
+--- script logic
+
+local GLOBAL_KEY = 'immortal-cravings'
+
+enabled = enabled or false
+function isEnabled()
+    return enabled
+end
+
+local function persist_state()
+    dfhack.persistent.saveSiteData(GLOBAL_KEY, {
+        enabled=enabled,
+    })
+end
+
+--- Load the saved state of the script
+local function load_state()
+    -- load persistent data
+    local persisted_data = dfhack.persistent.getSiteData(GLOBAL_KEY, {})
+    enabled = persisted_data.enabled or false
+end
+
+DrinkAlcohol = df.need_type['DrinkAlcohol']
+EatGoodMeal = df.need_type['EatGoodMeal']
+
+---@type integer[]
+watched = {}
+
+threshold = -9000
+
+---unit loop: check for idle watched units and create eat/drink jobs for them
+local function unit_loop()
+    -- print(('immortal-cravings: running unit loop (%d watched units)'):format(#watched))
+    ---@type integer[]
+    local kept = {}
+    for _, unit_id in ipairs(watched) do
+        local unit = df.unit.find(unit_id)
+        if unit and not (unit.flags1.caged or unit.flags1.chained) then
+            if not idle.unitIsAvailable(unit) then
+                table.insert(kept, unit.id)
+            else
+                --
+                for _, need in ipairs(unit.status.current_soul.personality.needs) do
+                    if need.id == DrinkAlcohol and need.focus_level < threshold then
+                        goDrink(unit)
+                        goto next_unit
+                    elseif need.id == EatGoodMeal and need.focus_level < threshold then
+                        goEat(unit)
+                        goto next_unit
+                    end
+                end
+            end
+        else
+            -- print('immortal-cravings: unit gone or caged')
+        end
+        ::next_unit::
+    end
+    watched = kept
+    if #watched == 0 then
+        -- print('immortal-cravings: no more watched units, cancelling unit loop')
+        repeatutil.cancel(GLOBAL_KEY .. '-unit')
+    end
+end
+
+---main loop: look for citizens with personality needs for food/drink but w/o physiological need
+local function main_loop()
+    print('immortal-cravings watching:')
```

Comment:
this prints in this function are probably too spammy. the prints for when actions are actually taken is enough